src/Microsoft.ML.Data/StaticPipe/TrainerEstimatorReconciler.cs

```diff
@@ -0,0 +1,335 @@
+            {
+                Contracts.CheckValue(estimatorMaker, nameof(estimatorMaker));
+                _estMaker = estimatorMaker;
+                Contracts.Assert(_inputs.Length == 2 || _inputs.Length == 3);
+
+                Output = (new Impl(this), new Impl(this), new ImplBool(this));
+            }
+
+            private static PipelineColumn[] MakeInputs(Scalar<bool> label, Vector<float> features, Scalar<float> weights)
+                => weights == null ? new PipelineColumn[] { label, features } : new PipelineColumn[] { label, features, weights };
+
+            protected override IEstimator<ITransformer> ReconcileCore(IHostEnvironment env, string[] inputNames)
+            {
+                Contracts.AssertValue(env);
+                env.Assert(Utils.Size(inputNames) == _inputs.Length);
+                return _estMaker(env, inputNames[0], inputNames[1], inputNames.Length > 2 ? inputNames[2] : null);
+            }
+
+            private sealed class Impl : Scalar<float>
+            {
+                public Impl(BinaryClassifier rec) : base(rec, rec._inputs) { }
+            }
+
+            private sealed class ImplBool : Scalar<bool>
+            {
+                public ImplBool(BinaryClassifier rec) : base(rec, rec._inputs) { }
+            }
+        }
+
+        /// <summary>
+        /// A reconciler for regression capable of handling the most common cases for binary classification
+        /// that does not necessarily have calibrated outputs.
+        /// </summary>
+        public sealed class BinaryClassifierNoCalibration : TrainerEstimatorReconciler
+        {
+            /// <summary>
+            /// The delegate to parameterize the <see cref="BinaryClassifier"/> instance.
+            /// </summary>
+            /// <param name="env">The environment with which to create the estimator</param>
+            /// <param name="label">The label column name</param>
+            /// <param name="features">The features column name</param>
+            /// <param name="weights">The weights column name, or <c>null</c> if the reconciler was constructed with <c>null</c> weights</param>
+            /// <returns></returns>
+            public delegate IEstimator<ITransformer> EstimatorMaker(IHostEnvironment env, string label, string features, string weights);
+
+            private readonly EstimatorMaker _estMaker;
+            private static readonly string[] _fixedOutputNamesProb = new[] { DefaultColumnNames.Score, DefaultColumnNames.Probability, DefaultColumnNames.PredictedLabel };
+            private static readonly string[] _fixedOutputNames = new[] { DefaultColumnNames.Score, DefaultColumnNames.PredictedLabel };
+
+            /// <summary>
+            /// The general output for binary classifiers.
+            /// </summary>
+            public (Scalar<float> score, Scalar<bool> predictedLabel) Output { get; }
+
+            protected override IEnumerable<PipelineColumn> Outputs { get; }
```

Comment:
I'm not sure I fully grasp the difference between `Output` and `Outputs`...

Reply:
I'll add a comment. They only really differ in content in this one class, due to the fact that are compile time we can only be sure two are present, while at runtime we also have to be sensitive as to whether we are in fact producing that extra probability column.

---
In reply to: [215371730](https://github.com/dotnet/machinelearning/pull/837#discussion_r215371730) [](ancestors = 215371730)